### PR TITLE
Push Notifications

### PR DESCRIPTION
**Pull request by dwieeb**
TODO:
- [x] everything
- [x] cli config types (example: https://github.com/ionic-team/capacitor-plugins/pull/94/files#diff-60a5e322448a3b93ce1e805ae5a11dba89030ef6ebc775a2487ae3fb9a41a8caR1-R43)

### CODE REVIEW

File: push-notifications/android/build.gradle [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    }
+    dependencies {
+        classpath 'com.android.tools.build:gradle:4.1.1'
+    }
+}
+
+apply plugin: 'com.android.library'
+
+android {
+    compileSdkVersion project.hasProperty('compileSdkVersion') ? rootProject.ext.compileSdkVersion : 29
```
Comment (dwieeb) [inline push-notifications/android/build.gradle:22]:
```suggestion
    compileSdkVersion project.hasProperty('compileSdkVersion') ? rootProject.ext.compileSdkVersion : 30
```

File: push-notifications/android/build.gradle [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    defaultConfig {
+        minSdkVersion project.hasProperty('minSdkVersion') ? rootProject.ext.minSdkVersion : 21
+        targetSdkVersion project.hasProperty('targetSdkVersion') ? rootProject.ext.targetSdkVersion : 29
```
Comment (dwieeb) [inline push-notifications/android/build.gradle:25]:
```suggestion
        targetSdkVersion project.hasProperty('targetSdkVersion') ? rootProject.ext.targetSdkVersion : 30
```

File: push-notifications/android/build.gradle [hunk 3]
```diff
@@ -0,0 +1,60 @@
+ext {
+    junitVersion =  project.hasProperty('junitVersion') ? rootProject.ext.junitVersion : '4.12'
+    androidxAppCompatVersion = project.hasProperty('androidxAppCompatVersion') ? rootProject.ext.androidxAppCompatVersion : '1.1.0'
+    androidxJunitVersion =  project.hasProperty('androidxJunitVersion') ? rootProject.ext.androidxJunitVersion : '1.1.1'
+    androidxEspressoCoreVersion =  project.hasProperty('androidxEspressoCoreVersion') ? rootProject.ext.androidxEspressoCoreVersion : '3.2.0'
+    firebaseMessagingVersion =  project.hasProperty('firebaseMessagingVersion') ? rootProject.ext.firebaseMessagingVersion : '20.1.2'
```
Comment (dwieeb) [inline push-notifications/android/build.gradle:6]:
```suggestion
    junitVersion = project.hasProperty('junitVersion') ? rootProject.ext.junitVersion : '4.12'
    androidxAppCompatVersion = project.hasProperty('androidxAppCompatVersion') ? rootProject.ext.androidxAppCompatVersion : '1.1.0'
    androidxJunitVersion = project.hasProperty('androidxJunitVersion') ? rootProject.ext.androidxJunitVersion : '1.1.1'
    androidxEspressoCoreVersion = project.hasProperty('androidxEspressoCoreVersion') ? rootProject.ext.androidxEspressoCoreVersion : '3.2.0'
    firebaseMessagingVersion = project.hasProperty('firebaseMessagingVersion') ? rootProject.ext.firebaseMessagingVersion : '20.1.2'
```

File: push-notifications/ios/Plugin/PushNotifications.swift [hunk 4]
```diff
@@ -0,0 +1,7 @@
+import Foundation
+
+@objc public class PushNotifications: NSObject {
+    @objc public func echo(_ value: String) -> String {
+        return value
+    }
+}
```
Comment (dwieeb) [inline push-notifications/ios/Plugin/PushNotifications.swift:7]:
```suggestion
```

File: push-notifications/ios/Plugin/PushNotificationsPlugin.swift [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+import FirebaseCore
+import FirebaseMessaging
+import UserNotifications
+
+enum PushNotificationError: Error {
+    case tokenParsingFailed
+    case tokenRegistrationFailed
+}
+
+enum LocationPermissions: String {
```
Comment (dwieeb) [inline push-notifications/ios/Plugin/PushNotificationsPlugin.swift:12]: Should probably be `PushNotificationsPermissions`

File: push-notifications/ios/Plugin/PushNotificationsPlugin.swift [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+            var result: LocationPermissions = .prompt
+
+            switch status {
+            case .notDetermined:
+                result = .prompt
+            case .denied:
+                result = .denied
+            case .ephemeral, .authorized, .provisional:
+                result = .granted
+            default:
```
Comment (dwieeb) [inline push-notifications/ios/Plugin/PushNotificationsPlugin.swift:88]: I've seen `@unknown default:` in other plugins for these which will trigger a compile warning if a new enum is added, I think. Do we want to do that here?

File: push-notifications/package.json [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    "@capacitor/android": "^3.0.0-alpha.7",
+    "@capacitor/core": "^3.0.0-alpha.7",
+    "@capacitor/docgen": "^0.0.10",
+    "@capacitor/ios": "^3.0.0-alpha.7",
+    "@ionic/eslint-config": "^0.3.0",
+    "@ionic/prettier-config": "^1.0.1",
+    "@ionic/swiftlint-config": "^1.1.2",
+    "@types/jest": "^26.0.15",
+    "eslint": "^7.11.0",
+    "jest": "^26.6.3",
```
Comment (dwieeb) [inline push-notifications/package.json:56]: We should remove jest and related packages/config/scripts from this plugin for the time being (take a look at other `package.json` files)

File: push-notifications/package.json [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+      "src": "ios"
+    },
+    "android": {
+      "src": "android"
+    }
+  },
+  "jest": {
+    "preset": "ts-jest"
+  }
+}
```
Comment (dwieeb) [inline push-notifications/package.json:84]: add `publishConfig` (see other `package.json` files)

File: push-notifications/ios/Plugin/PushNotificationsPlugin.swift (hunk 5 above)
Comment (theproducer) [reply push-notifications/ios/Plugin/PushNotificationsPlugin.swift:12]: Good catch!

File: push-notifications/package.json [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+  "keywords": [
+    "capacitor",
+    "plugin",
+    "native"
+  ],
+  "scripts": {
+    "verify": "npm run verify:ios && npm run verify:android && npm run verify:web",
+    "verify:ios": "cd ios && pod install && xcodebuild -workspace Plugin.xcworkspace -scheme Plugin && cd ..",
+    "verify:android": "cd android && ./gradlew clean build test && cd ..",
+    "verify:web": "npm run build && npm test",
```
Comment (dwieeb) [inline push-notifications/package.json:33]:
```suggestion
    "verify:web": "npm run build",
```

**Review by dwieeb — COMMENTED** [review]
We need to switch from `Color.parseColor()` to `WebColor.parseColor()` (see https://github.com/ionic-team/capacitor-plugins/pull/140 and https://github.com/ionic-team/capacitor/pull/3947)

The docs for `lightColor` will also need to be updated

**Review by carlpoole — APPROVED** [review]

File: push-notifications/src/definitions.ts [hunk 10]
```diff
@@ -0,0 +1,330 @@
...
+   */
+  lights?: boolean;
+
+  /**
+   * The light color for notifications posted to this channel.
+   *
+   * Only supported if lights are enabled on this channel and the device
+   * supports it.
+   *
+   * Supported color formats are `#RRGGBB` and `#AARRGGBB`.
```
Comment (dwieeb) [inline push-notifications/src/definitions.ts:304]:
```suggestion
   * Supported color formats are `#RRGGBB` and `#RRGGBBAA`.
```

**Comment by dwieeb** [comment]
LGTM 💙

**Review by theproducer — APPROVED** [review]

File: push-notifications/ios/Plugin/PushNotificationsDelegate.swift [hunk 11]
```diff
@@ -0,0 +1,91 @@
+import Capacitor
+import UserNotifications
+
+public class PushNotificationsDelegate: NSObject, NotificationHandlerProtocol {
+    public var plugin: CAPPlugin?
```
Comment (ikeith) [inline push-notifications/ios/Plugin/PushNotificationsDelegate.swift:5]: This creates a retain cycle that will cause problems.
```suggestion
    public weak var plugin: CAPPlugin?
```

File: push-notifications/ios/Plugin/PushNotificationsDelegate.swift [hunk 12]
```diff
@@ -0,0 +1,91 @@
+import Capacitor
+import UserNotifications
+
+public class PushNotificationsDelegate: NSObject, NotificationHandlerProtocol {
```
Comment (ikeith) [inline push-notifications/ios/Plugin/PushNotificationsDelegate.swift:4]: I think this class name should be revisited. The term `delegate` has a lot of connotations which don't necessarily apply and can cause confusion. The protocol it adopts is a `handler` protocol not a `delegate` protocol, after all. `PushNotificationHandler` perhaps?

File: push-notifications/ios/Plugin/PushNotificationsPlugin.swift [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+
+    override public func load() {
+        FirebaseApp.configure()
+        self.bridge?.notificationRouter.pushNotificationHandler = self.notificationDelegateHandler
+        self.notificationDelegateHandler.plugin = self
+
+        NotificationCenter.default.addObserver(self,
+                                               selector: #selector(self.didRegisterForRemoteNotificationsWithDeviceToken(notification:)),
+                                               name: Notification.Name(Notification.Name.capacitorDidRegisterForRemoteNotifications.self.rawValue),
+                                               object: nil)
```
Comment (ikeith) [inline push-notifications/ios/Plugin/PushNotificationsPlugin.swift:31]:
```suggestion
        NotificationCenter.default.addObserver(self,
                                               selector: #selector(self.didRegisterForRemoteNotificationsWithDeviceToken(notification:)),
                                               name: .capacitorDidRegisterForRemoteNotifications,
                                               object: nil)
```

File: push-notifications/ios/Plugin/PushNotificationsPlugin.swift [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+
+        NotificationCenter.default.addObserver(self,
+                                               selector: #selector(self.didFailToRegisterForRemoteNotificationsWithError(notification:)),
+                                               name: Notification.Name(Notification.Name.capacitorDidFailToRegisterForRemoteNotifications.self.rawValue),
+                                               object: nil)
```
Comment (ikeith) [inline push-notifications/ios/Plugin/PushNotificationsPlugin.swift:36]:
```suggestion
        NotificationCenter.default.addObserver(self,
                                               selector: #selector(self.didFailToRegisterForRemoteNotificationsWithError(notification:)),
                                               name: .capacitorDidFailToRegisterForRemoteNotifications,
                                               object: nil)
```

File: push-notifications/ios/Plugin/PushNotificationsPlugin.swift [hunk 15]
```diff
@@ -0,0 +1,188 @@
+import Foundation
+import Capacitor
+import FirebaseCore
```
Comment (ikeith) [inline push-notifications/ios/Plugin/PushNotificationsPlugin.swift:3]: Why is this being tied to firebase? Firebase may be required on Android but iOS should service-agnostic about where the push notifications are coming from. Adding that dependency has significant implications.

**Review by ikeith — CHANGES_REQUESTED** [review]

**Review by ikeith — APPROVED** [review]

**Pull request merged**